### ISSUE TITLE
Blank Chapters

### ISSUE THREAD

**Issue by BalkrishanSingh** [dipu-bd/lightnovel-crawler#1645]
## Describe the bug
EPUB having a few blank chapters when downloading Mmorpg Rise Of The Peerless Pumpkinmancer  from all the sources available when queried 

I have attached the compressed epub from one of my attempts to download the book. 
[Mmorpg Rise Of The Peerless Pumpkinmancer c1-414.epub.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9653911/Mmorpg.Rise.Of.The.Peerless.Pumpkinmancer.c1-414.epub.zip)

I used this source in particular for that attempt 
https://www.novelupdates.cc/MMORPG-Rise-of-the-Peerless-Pumpkinmancer/

These were all the sources that were available to me when i queried. I have tried out most of them and the results were the same no matter what.
 1. http://novelfull.live/novel/mmorpg-rise-of-the-peerless-pumpkinmancer
      💁 Latest chapter: Chapter 367: Gang VS Seaweed Monster!
   2. https://1stkissnovel.love/novel/mmorpg-rise-of-the-peerless-pumpkinmancer/
      💁 Chapter 320 Summer Heat! | Rating: 0
   3. https://allnovel.org/mmorpg-rise-of-the-peerless-pumpkinmancer.html
      💁 Chapter 367: Gang VS Seaweed Monster!
   4. https://allnovelfull.com/mmorpg-rise-of-the-peerless-pumpkinmancer.html
      💁 Chapter 367: Gang VS Seaweed Monster!
   5. https://hotnovelfull.com/n/mmorpg-rise-of-the-peerless-pumpkinmancer-novel-full
      💁 last chapter : Chapter 367: Gang VS Seaweed Monster!
   6. https://latestnovel.net/novel/mmorpg-rise-of-the-peerless-pumpkinmancer/
      💁 Chapter 237
   7. https://m.readlightnovel.cc/MMORPG：-Rise-of-the-Peerless-Pumpkinmancer/
      💁 ClasslessAscension
   8. https://www.ltnovel.com/novel/mmorpg-rise-of-the-peerless-pumpkinmancer.html
      💁 362 Chapters | 1 days ago | Status: Ongoing
   9. https://www.novelmt.com/novel/mmorpg-rise-of-the-peerless-pumpkinmancer.html
      💁 Chapters | 1 hours ago | Status: Ongoing
   10. https://www.novelupdates.cc/MMORPG：-Rise-of-the-Peerless-Pumpkinmancer/
      💁 Game | Rating: 8.0
   11. https://www.readwn.com/novel/mmorpg-rise-of-the-peerless-pumpkinmancer.html
      💁 Chapters | 1 hours ago | Status: Ongoing
   12. https://www.webnovel.com/book/21417395905200705
      💁 Video Games | Score: 4.7
## Let us know

**App source**:   PIP  
**App version**:   2.34.0  
**Your OS**:   Mac

**Comment by idMysteries** [comment]
Even in webnovel.com empty chapters?

**Comment by BalkrishanSingh** [comment]
Yup. I had't tried it out earlier because i could only get 40 chapters but there is empty chapters from webnovel.com as well
[Mmorpg Rise Of The Peerless Pumpkinmancer c1-40.epub.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9654248/Mmorpg.Rise.Of.The.Peerless.Pumpkinmancer.c1-40.epub.zip)

**Comment by idMysteries** [comment]
I tested it on version 3.0 and I don't have this bug.
@dipu-bd

**Comment by BalkrishanSingh** [comment]
Oh am i on a outdated version? I had done pip install -U Lightnovel-crawler to install it. How should i update it?

**Comment by idMysteries** [comment]
Perhaps the problem was in html minify @dipu-bd 

The new version is only in the dev branch.

**Comment by idMysteries** [comment]
https://github.com/dipu-bd/lightnovel-crawler/archive/refs/heads/dev.zip

**Comment by idMysteries** [comment]
```
pip install -r requirements.txt

python lncrawl
```

**Comment by BalkrishanSingh** [comment]
Tried downloading using version 3.0 and i still seem to be having the same issue.
Source : https://latestnovel.net/novel/mmorpg-rise-of-the-peerless-pumpkinmancer/
EPUB :  [Mmorpg Rise Of The Peerless Pumpkinmancer c1-237.epub.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9654397/Mmorpg.Rise.Of.The.Peerless.Pumpkinmancer.c1-237.epub.zip)

**Comment by BalkrishanSingh** [comment]
<img width="1358" alt="Screenshot 2022-09-27 at 15 38 17" src="https://user-images.githubusercontent.com/93041640/192498317-cbff24cc-18a9-40a8-8961-5b6858be6f78.png">
A Side issue but it seems to be giving me an error whenever i try to open the folder where the file is downloaded on the version 3.0

**Comment by idMysteries** [comment]
Can you download epub and json formats please? Give me json zip

**Comment by BalkrishanSingh** [comment]
Sure. Also tell me if you want the downloads from a specific source. 
The one i am sending now is from novelupdate
Here it is : [Mmorpg Rise Of The Peerless Pumpkinmancer.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9654463/Mmorpg.Rise.Of.The.Peerless.Pumpkinmancer.zip)

**Comment by idMysteries** [comment]
@dipu-bd the json file is the same as mine. But it is converted to epub incorrectly. Mac issue?

**Comment by idMysteries** [comment]
Is there the same problem with fb2?
If you're reading from an e-book, try a format other than epub.

**Comment by idMysteries** [comment]
@dipu-bd I checked everything I could, everything works for me, but there is no Mac at hand.

**Comment by BalkrishanSingh** [comment]
Hmm i am having trouble downloadings formats other than EPUB and Json as it requires Calibre and even tho I have installed it. lncrawler doesn't seem to be recognising it as it still refuses to download other formats

**Comment by idMysteries** [comment]
Yes, I realized that epub is converted to other formats.

The Web format should work.

**Comment by BalkrishanSingh** [comment]
Okay so Web,Text and JSOn formats seem to have no issue. I tried downloading it from novelupdates again on version 2.34
[Mmorpg Rise Of The Peerless Pumpkinmancer.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9654772/Mmorpg.Rise.Of.The.Peerless.Pumpkinmancer.zip)

**Comment by idMysteries** [comment]
The problem is ebooklib on mac. I do not know how to fix it.

**Comment by idMysteries** [comment]
What is your main language on Mac OS? English?

**Comment by BalkrishanSingh** [comment]
Yeah. It is English

**Comment by BalkrishanSingh** [comment]
Hmm It seems to be fine now.
Epub doesn't have blank chapters anymore.
The only thing i have changed would be the source which is https://hotnovelfull.com/n/mmorpg-rise-of-the-peerless-pumpkinmancer-novel-full and my system language since i am from india and the main language on my mac is English(India) and I changed it to English(US)
[Mmorpg Rise Of The Peerless Pumpkinmancer.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9654956/Mmorpg.Rise.Of.The.Peerless.Pumpkinmancer.zip)

**Comment by BalkrishanSingh** [comment]
Seems to be related to source since when i changed the source to novelupdates. the issue is still there

**Comment by idMysteries** [comment]
This is not a source problem, because it works for me. This is an ebooklib issue.

**Comment by idMysteries** [comment]
[epub.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9655203/epub.zip)

**Comment by BalkrishanSingh** [comment]
Thanks.

**Comment by idMysteries** [comment]
Say the mac version please

**Comment by BalkrishanSingh** [comment]
macOS 12.2.1 21D62 arm64

**Comment by idMysteries** [comment]
> Perhaps the problem was in html minify @dipu-bd
> 
> The new version is only in the dev branch.

minify-html doesn't support arm64 macOs
https://github.com/wilsonzlin/minify-html/commit/29e7b1c18705cbb534c2d8cabe11f66731524ac0

**Comment by idMysteries** [comment]
@dipu-bd wait... min python 3.8???

Package: [minify-html](https://pypi.org/project/minify-html)
Binding: [PyO3](https://github.com/PyO3/pyo3)
Platforms: Linux (ARM64 and x64), macOS (x64), Windows (x64); Python 3.8 to 3.10

**Comment by BalkrishanSingh** [comment]
Oh btw i am on Python 3.10.6

**Comment by dipu-bd** [comment]
Install latest changes from `dev` branch and check if the issue is solved there:

```
$ pip install -U https://github.com/dipu-bd/lightnovel-crawler/tarball/refs/heads/dev#egg=lightnovel-crawler
```

**Comment by idMysteries** [comment]
He tested dev branch @dipu-bd

**Comment by dipu-bd** [comment]
@idMysteries @BalkrishanSingh can I safely close this issue?

**Comment by dipu-bd** [comment]
Can someone summarize, what is the issue pending to be solved in this ticket? I actually lost track.

**Comment by idMysteries** [comment]
I think

> minify-html doesn't support arm64 macOs
https://github.com/wilsonzlin/minify-html/commit/29e7b1c18705cbb534c2d8cabe11f66731524ac0

**Comment by idMysteries** [comment]
But there was a problem before that update.

**Comment by idMysteries** [comment]
I think the problem may be in ebooklib on mac (arm64)

**Comment by dipu-bd** [comment]
There is good chance that minify-html will cause problem. minify-html is actually using a rust app to minify. I have yet to find any better alternatives in python.

**Comment by idMysteries** [comment]
Everything works for me, but on the mac epub it turns out with broken chapters.

**Comment by dipu-bd** [comment]
> I think the problem may be in ebooklib on mac (arm64)

I lost track about which problem you are discussing actually. Do you mean that the chapters are blank, even when html source is available on mac (arm64)?

**Comment by idMysteries** [comment]
Yes... Not empty, but broken.

**Comment by dipu-bd** [comment]
Okay, I understand now. It might be a problem of minify-html. Because, ebooklib does nothing but compress the html files.

**Comment by idMysteries** [comment]
[epubs.zip](https://github.com/dipu-bd/lightnovel-crawler/files/9658946/epubs.zip)

**Comment by idMysteries** [comment]
> Okay, I understand now. It might be a problem of minify-html. Because, ebooklib does nothing but compress the html files.

it broken on version 2.34 also

**Comment by idMysteries** [comment]
minify-html on version 3.0?

**Comment by idMysteries** [comment]
By the way, he has a working download in text, json, web without broken chapters.
So I think this is not a minify-html problem.
I am 67% sure that the problem is in ebooklib

**Comment by idMysteries** [comment]
WEB without broken chapters!!!! Shit...  is not a minify-html problem

**Comment by idMysteries** [comment]
The web format definitely uses minify-html. But there are no broken chapters!

**Comment by idMysteries** [comment]
ebooklib sucks!

**Comment by dipu-bd** [comment]
@BalkrishanSingh Please check if this issue still exists in the recently released v3.0?
